File: 70-tools/gftd/gftd/projector.go
// projector.go — gftd projector: project lifecycle + blocker tracking CLI.
//
// Uses the projector.* MCP tools / XRPC methods backed by:
//   - vertex_project_props (lifecycle_state, progress_permille)
//   - vertex_projector_blocker (per-project blockers)
//   - mv_projector_project_status (aggregated status MV)
//   - LangGraph projector_lifecycle graph (state transitions)
//   - Pregel BSP blocker propagation (dependency graph)
//
// Examples:
//
//	gftd projector create --name "Vultr A16 GPU deployment" --org default
//	gftd projector status --id proj:1234:abc
//	gftd projector update --id proj:1234:abc --progress 500 --state active
//	gftd projector blocker add --project proj:1234:abc --title "Vultr product gate" --type system
//	gftd projector blocker resolve --id blk:5678:def
//	gftd projector list --state blocked
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

const projectorMCPBase = "https://atproto.etzhayyim.com"

func runProjector(args []string) error {
	if len(args) == 0 {
		return fmt.Errorf("usage: gftd projector <create|status|update|blocker|list> [flags]")
	}
	switch args[0] {
	case "create":
		return runProjectorCreate(args[1:])
	case "status", "get":
		return runProjectorStatus(args[1:])
	case "update":
		return runProjectorUpdate(args[1:])
	case "blocker":
		if len(args) < 2 {
			return fmt.Errorf("usage: gftd projector blocker <add|resolve> [flags]")
		}
		switch args[1] {
		case "add":
			return runProjectorBlockerAdd(args[2:])
		case "resolve":
			return runProjectorBlockerResolve(args[2:])
		default:
			return fmt.Errorf("unknown blocker subcommand: %s", args[1])
		}
	case "list":
		return runProjectorList(args[1:])
	default:
		return fmt.Errorf("unknown projector subcommand: %s\navailable: create, status, update, blocker add/resolve, list", args[0])
	}
}

func projectorMCPCall(toolName string, arguments map[string]any) (map[string]any, error) {
	payload := map[string]any{
		"jsonrpc": "2.0",
		"id":      1,
		"method":  "tools/call",
		"params": map[string]any{
			"name":      toolName,
			"arguments": arguments,
		},
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", projectorMCPBase+"/mcp", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if tok := os.Getenv("GFTD_AGENT_TOKEN"); tok != "" {
		req.Header.Set("Authorization", "Bearer "+tok)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("MCP request failed: %w", err)
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var rpcResp map[string]any
	if err := json.Unmarshal(raw, &rpcResp); err != nil {
		return nil, fmt.Errorf("response parse error: %w\nraw: %s", err, string(raw))
	}

	if errVal, ok := rpcResp["error"]; ok {
		return nil, fmt.Errorf("MCP error: %v", errVal)
	}

	result, _ := rpcResp["result"].(map[string]any)
	if result == nil {
		return rpcResp, nil
	}

	// Unwrap content[0].text if present
	if content, ok := result["content"].([]any); ok && len(content) > 0 {
		if first, ok := content[0].(map[string]any); ok {
			if text, ok := first["text"].(string); ok {
				var inner map[string]any
				if json.Unmarshal([]byte(text), &inner) == nil {
					return inner, nil
				}
			}
		}
	}
	return result, nil
}

func printProjectorResult(result map[string]any) {
	out, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(out))
}

// ── create ──────────────────────────────────────────────────────────────────

func runProjectorCreate(args []string) error {
	fs := flag.NewFlagSet("projector create", flag.ContinueOnError)
	name := fs.String("name", "", "Project name (required)")
	description := fs.String("desc", "", "Project description")
	orgID := fs.String("org", "default", "org_id")
	parentID := fs.String("parent", "", "Parent project vertex_id")
	targetDate := fs.String("due", "", "Target date ISO 8601 (e.g. 2026-06-30)")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *name == "" {
		return fmt.Errorf("--name is required")
	}

	arguments := map[string]any{"name": *name, "orgId": *orgID}
	if *description != "" {
		arguments["description"] = *description
	}
	if *parentID != "" {
		arguments["parentId"] = *parentID
	}
	if *targetDate != "" {
		arguments["targetDate"] = *targetDate
	}

	result, err := projectorMCPCall("projector.create_project", arguments)
	if err != nil {
		return err
	}
	printProjectorResult(result)
	return nil
}

// ── status ───────────────────────────────────────────────────────────────────

func runProjectorStatus(args []string) error {
	fs := flag.NewFlagSet("projector status", flag.ContinueOnError)
	id := fs.String("id", "", "Project vertex_id (required)")
	summarize := fs.Bool("summarize", false, "Include LLM summary")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *id == "" && fs.NArg() > 0 {
		*id = fs.Arg(0)
	}
	if *id == "" {
		return fmt.Errorf("--id is required")
	}

	result, err := projectorMCPCall("projector.get_status", map[string]any{
		"projectId": *id,
		"summarize": *summarize,
	})
	if err != nil {
		return err
	}

	// Pretty-print key fields
	fmt.Printf("Project:  %s\n", result["name"])
	fmt.Printf("State:    %s\n", result["lifecycleState"])
	if pp, ok := result["progressPermille"].(float64); ok {
		fmt.Printf("Progress: %.1f%%\n", pp/10.0)
	}
	if obc, ok := result["openBlockerCount"].(float64); ok && obc > 0 {
		fmt.Printf("Blockers: %.0f open\n", obc)
	}
	if td, ok := result["targetDate"].(string); ok && td != "" {
		fmt.Printf("Due:      %s\n", td)
	}
	if sum, ok := result["summary"].(string); ok && sum != "" {
		fmt.Printf("Summary:  %s\n", sum)
	}
	return nil
}

// ── update ───────────────────────────────────────────────────────────────────

func runProjectorUpdate(args []string) error {
	fs := flag.NewFlagSet("projector update", flag.ContinueOnError)
	id := fs.String("id", "", "Project vertex_id (required)")
	progress := fs.Int("progress", -1, "Progress permille 0-1000")
	state := fs.String("state", "", "lifecycle_state: planning|active|blocked|done")
	targetDate := fs.String("due", "", "Target date ISO 8601")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *id == "" {
		return fmt.Errorf("--id is required")
	}

	arguments := map[string]any{"projectId": *id}
	if *progress >= 0 {
		arguments["progressPermille"] = *progress
	}
	if *state != "" {
		arguments["lifecycleState"] = *state
	}
	if *targetDate != "" {
		arguments["targetDate"] = *targetDate
	}

	result, err := projectorMCPCall("projector.update_status", arguments)
	if err != nil {
		return err
	}
	printProjectorResult(result)
	return nil
}

// ── blocker add ──────────────────────────────────────────────────────────────

func runProjectorBlockerAdd(args []string) error {
	fs := flag.NewFlagSet("projector blocker add", flag.ContinueOnError)
	project := fs.String("project", "", "Project vertex_id (required)")
	title := fs.String("title", "", "Blocker title (required)")
	description := fs.String("desc", "", "Blocker description")
	blockerType := fs.String("type", "technical", "Blocker type: financial|personnel|approval|system|external|technical")
	severity := fs.String("severity", "medium", "Severity: low|medium|high|critical")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *project == "" || *title == "" {
		return fmt.Errorf("--project and --title are required")
	}

	arguments := map[string]any{
		"projectId":   *project,
		"title":       *title,
		"blockerType": *blockerType,
		"severity":    *severity,
	}
	if *description != "" {
		arguments["description"] = *description
	}

	result, err := projectorMCPCall("projector.add_blocker", arguments)
	if err != nil {
		return err
	}
	printProjectorResult(result)
	return nil
}

// ── blocker resolve ───────────────────────────────────────────────────────────

func runProjectorBlockerResolve(args []string) error {
	fs := flag.NewFlagSet("projector blocker resolve", flag.ContinueOnError)
	id := fs.String("id", "", "Blocker vertex_id (required)")
	resolution := fs.String("resolution", "", "How the blocker was resolved")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *id == "" && fs.NArg() > 0 {
		*id = fs.Arg(0)
	}
	if *id == "" {
		return fmt.Errorf("--id is required")
	}

	arguments := map[string]any{"blockerId": *id}
	if *resolution != "" {
		arguments["resolution"] = *resolution
	}

	result, err := projectorMCPCall("projector.resolve_blocker", arguments)
	if err != nil {
		return err
	}
	printProjectorResult(result)
	return nil
}

// ── list ──────────────────────────────────────────────────────────────────────

func runProjectorList(args []string) error {
	fs := flag.NewFlagSet("projector list", flag.ContinueOnError)
	orgID := fs.String("org", "", "Filter by org_id")
	state := fs.String("state", "", "Filter by lifecycle_state")
	limit := fs.Int("limit", 50, "Max results (1-100)")
	jsonOut := fs.Bool("json", false, "Output raw JSON")
	if err := fs.Parse(args); err != nil {
		return err
	}

	arguments := map[string]any{"limit": *limit}
	if *orgID != "" {
		arguments["orgId"] = *orgID
	}
	if *state != "" {
		arguments["lifecycleState"] = *state
	}

	result, err := projectorMCPCall("projector.list_projects", arguments)
	if err != nil {
		return err
	}

	if *jsonOut {
		printProjectorResult(result)
		return nil
	}

	projects, _ := result["projects"].([]any)
	if len(projects) == 0 {
		fmt.Println("no projects found")
		return nil
	}

	fmt.Printf("%-12s %-10s %8s %8s  %s\n", "STATE", "PROGRESS", "BLOCKERS", "DUE", "NAME")
	fmt.Println(strings.Repeat("-", 72))
	for _, p := range projects {
		proj, ok := p.(map[string]any)
		if !ok {
			continue
		}
		st, _ := proj["lifecycleState"].(string)
		name, _ := proj["name"].(string)
		pp, _ := proj["progressPermille"].(float64)
		obc, _ := proj["openBlockerCount"].(float64)
		td, _ := proj["targetDate"].(string)
		if td == "" {
			td = "-"
		}
		fmt.Printf("%-12s %7.1f%% %8.0f  %8s  %s\n", st, pp/10.0, obc, td, name)
	}
	return nil
}

File: 70-tools/etzhayyim-py/src/etzhayyim/projector.py
"""projector subcommands — Python port of Go etzhayyim/projector.go (ADR-2605151500).

Calls the projector MCP tool via JSON-RPC 2.0 over HTTPS.
Auth: etzhayyim_AGENT_TOKEN env var → Authorization: Bearer {token}.
MCP endpoint: POST {pds}/mcp
"""

from __future__ import annotations

import json
import os
import sys

import click
import httpx

from .auth import resolve_pds

_MCP_METHOD = "tools/call"
_JSONRPC = "2.0"
_REQ_ID = 1


def _agent_token() -> str | None:
    return os.environ.get("etzhayyim_AGENT_TOKEN")


def _mcp_headers() -> dict[str, str]:
    headers: dict[str, str] = {"Content-Type": "application/json"}
    if tok := _agent_token():
        headers["Authorization"] = f"Bearer {tok}"
    return headers


def _mcp_call(tool_name: str, arguments: dict) -> dict:
    """POST JSON-RPC 2.0 tools/call to {pds}/mcp and return the parsed result dict."""
    pds = resolve_pds()
    url = f"{pds}/mcp"
    payload = {
        "jsonrpc": _JSONRPC,
        "id": _REQ_ID,
        "method": _MCP_METHOD,
        "params": {"name": tool_name, "arguments": arguments},
    }
    resp = httpx.post(url, json=payload, headers=_mcp_headers(), timeout=30)
    resp.raise_for_status()
    data = resp.json()
    if "error" in data:
        err = data["error"]
        raise click.ClickException(f"MCP error {err.get('code')}: {err.get('message')}")
    result = data.get("result", {})
    # Unwrap content[0].text (MCP tool response envelope)
    content = result.get("content", [])
    if content and isinstance(content[0], dict):
        text = content[0].get("text", "")
        try:
            return json.loads(text)
        except (json.JSONDecodeError, TypeError):
            return {"text": text}
    return result


def _print_json(data: dict) -> None:
    click.echo(json.dumps(data, ensure_ascii=False, indent=2))


# ── CLI group ──────────────────────────────────────────────────────────────────

@click.group("projector")
def projector():
    """Manage projects via the projector MCP tool."""


# ── create ─────────────────────────────────────────────────────────────────────

@projector.command("create")
@click.argument("name")
@click.option("--org-id", envvar="etzhayyim_ORG_ID", default=None, help="Organization DID")
@click.option("--description", "-d", default=None)
@click.option("--parent-id", default=None)
@click.option("--target-date", default=None, help="ISO date e.g. 2026-12-31")
def create(name: str, org_id: str | None, description: str | None, parent_id: str | None, target_date: str | None):
    """Create a new project."""
    args: dict = {"name": name}
    if org_id:
        args["orgId"] = org_id
    if description:
        args["description"] = description
    if parent_id:
        args["parentId"] = parent_id
    if target_date:
        args["targetDate"] = target_date
    try:
        result = _mcp_call("projector.create_project", args)
        _print_json(result)
    except httpx.HTTPError as exc:
        raise click.ClickException(str(exc)) from exc


# ── status / get ───────────────────────────────────────────────────────────────

@projector.command("status")
@click.argument("project_id")
@click.option("--summarize/--no-summarize", default=True)
def status(project_id: str, summarize: bool):
    """Get project status."""
    try:
        result = _mcp_call("projector.get_status", {"projectId": project_id, "summarize": summarize})
        _print_json(result)
    except httpx.HTTPError as exc:
        raise click.ClickException(str(exc)) from exc


@projector.command("get")
@click.argument("project_id")
@click.option("--summarize/--no-summarize", default=True)
@click.pass_context
def get(ctx: click.Context, project_id: str, summarize: bool):
    """Alias for status."""
    ctx.invoke(status, project_id=project_id, summarize=summarize)


# ── update ─────────────────────────────────────────────────────────────────────

@projector.command("update")
@click.argument("project_id")
@click.option("--progress", type=int, default=None, help="Progress permille (0-1000)")
@click.option("--state", default=None, help="Lifecycle state")
@click.option("--target-date", default=None)
def update(project_id: str, progress: int | None, state: str | None, target_date: str | None):
    """Update project status."""
    args: dict = {"projectId": project_id}
    if progress is not None:
        args["progressPermille"] = progress
    if state:
        args["lifecycleState"] = state
    if target_date:
        args["targetDate"] = target_date
    try:
        result = _mcp_call("projector.update_status", args)
        _print_json(result)
    except httpx.HTTPError as exc:
        raise click.ClickException(str(exc)) from exc


# ── list ───────────────────────────────────────────────────────────────────────

@projector.command("list")
@click.option("--org-id", envvar="etzhayyim_ORG_ID", default=None)
@click.option("--state", default=None, help="Filter by lifecycle state")
@click.option("--limit", default=20, show_default=True)
def list_projects(org_id: str | None, state: str | None, limit: int):
    """List projects."""
    args: dict = {"limit": limit}
    if org_id:
        args["orgId"] = org_id
    if state:
        args["lifecycleState"] = state
    try:
        result = _mcp_call("projector.list_projects", args)
        _print_json(result)
    except httpx.HTTPError as exc:
        raise click.ClickException(str(exc)) from exc


# ── blocker ────────────────────────────────────────────────────────────────────

@projector.group("blocker")
def blocker():
    """Manage project blockers."""


@blocker.command("add")
@click.argument("project_id")
@click.argument("title")
@click.option("--type", "blocker_type", default="technical", show_default=True,
              help="Blocker type: technical|resource|external|process")
@click.option("--severity", default="medium", show_default=True,
              help="Severity: low|medium|high|critical")
@click.option("--description", "-d", default=None)
def blocker_add(project_id: str, title: str, blocker_type: str, severity: str, description: str | None):
    """Add a blocker to a project."""
    args: dict = {"projectId": project_id, "title": title, "blockerType": blocker_type, "severity": severity}
    if description:
        args["description"] = description
    try:
        result = _mcp_call("projector.add_blocker", args)
        _print_json(result)
    except httpx.HTTPError as exc:
        raise click.ClickException(str(exc)) from exc


@blocker.command("resolve")
@click.argument("blocker_id")
@click.option("--resolution", "-r", default=None)
def blocker_resolve(blocker_id: str, resolution: str | None):
    """Resolve a blocker."""
    args: dict = {"blockerId": blocker_id}
    if resolution:
        args["resolution"] = resolution
    try:
        result = _mcp_call("projector.resolve_blocker", args)
        _print_json(result)
    except httpx.HTTPError as exc:
        raise click.ClickException(str(exc)) from exc
